### PR TITLE
[rust] Added support for creating empty dataset

### LINKED ISSUE #954: Support creating an empty table

### PR DESCRIPTION
**Pull request by trueutkarsh**
Closes #954 

Currently the test only checks schema properties.  Are there more scenarios/checks that I should cover in the test ?
I can't figure out what else you can do with empty dataset since you can't change schema dynamically or add new values because of it.

Please let me know your thoughts over this.

### CODE REVIEW

File: rust/src/dataset.rs [hunk 1]
```diff
@@ -1070,8 +1071,9 @@ mod tests {
         let test_dir = tempdir().unwrap();
         let test_uri = test_dir.path().to_str().unwrap();
         let mut reader: Box<dyn RecordBatchReader> = Box::new(RecordBatchBuffer::empty());
-        let result = Dataset::write(&mut reader, test_uri, None).await;
-        assert!(matches!(result.unwrap_err(), Error::EmptyDataset { .. }));
+        let result = Dataset::write(&mut reader, test_uri, None).await.unwrap();
```
Comment (eddyxu) [inline rust/src/dataset.rs:1074]: Could you also test that `count_rows()` and `to_batch`, and `add new rows` work?
Also make sure they work in Python?

File: rust/src/dataset.rs [hunk 2]
```diff
@@ -309,7 +309,8 @@ impl Dataset {
                 return Err(Error::from(batch.as_ref().unwrap_err()));
             }
         } else {
-            return Err(Error::EmptyDataset);
+            warn!("Dataset is empty, proceeding with empty schema");
+            schema = Schema::try_from(&ArrowSchema::empty())?;
```
Comment (eddyxu) [inline rust/src/dataset.rs:313]: When creating an empty dataset, users should provide a schema.

I'd imagine that the semantics will be very similar to SQL `CREATE TABLE`.

**Review by eddyxu — CHANGES_REQUESTED** [review]

File: rust/src/dataset.rs [hunk 3]
```diff
@@ -309,7 +309,8 @@ impl Dataset {
                 return Err(Error::from(batch.as_ref().unwrap_err()));
             }
         } else {
-            return Err(Error::EmptyDataset);
```
Comment (eddyxu) [inline rust/src/dataset.rs:312]: We might be able to remove this `EmptyDataset` as error.

**Comment by eddyxu** [comment]
@trueutkarsh Lemme know if you need anything else. We'd love to have this PR to get in soon. 

Thanks again for contribution.

File: rust/src/dataset.rs (hunk 2 above)
Comment (trueutkarsh) [reply rust/src/dataset.rs]: Yes indeed. Up until now we were extracting schema from the records but now that records might be absent in the case of empty dataset, schema must be provided or error will be thrown. This is what mostly this latest commit covers.

File: rust/src/dataset.rs (hunk 3 above)
Comment (trueutkarsh) [reply rust/src/dataset.rs:320]: Yeah, indeed. It's gone in latest version.

File: rust/src/dataset.rs (hunk 1 above)
Comment (trueutkarsh) [reply rust/src/dataset.rs:1087]: Hi Lei,
I've tested count_rows() and adding new rows. I couldn't find anything regarding to_batch(). Could you please point me in the right direction regarding it.

**Comment by trueutkarsh** [comment]
Hi @eddyxu @wjones127,
Let me briefly explain what I've tried to do this time.

RecordBatchBuffer implemented trait schema (RecordBatchReader) by extracting schema from the batches it stored. Now that we have to support empty datasets with schema as well, we need to make sure that RecordBatchBuffer must store the schema for that case. Hence I added a new optional field in RecordBatchBuffer of type Option<SchemaRef>. Now whenever schema function is called upon dataset, if dataset has records, schema would be extracted from them or else the optional schema would be resolved and returned. 
This resulted in minor changes in constructing empty RecordBatchBuffer or with batches where in first case you must specify schema and in latter you can optionally skip but good practice to specify. 

Please let me know if you find something unclear in the code or any thing else you'd like me cover which I missed. 

Some doctests have been failing to which I made no changes so I would inspect what's going on there.

Thanks

File: rust/src/dataset.rs (hunk 1 above)
Comment (eddyxu) [reply rust/src/dataset.rs:1087]: My bad, maybe just try [`to_table()`](https://lancedb.github.io/lance/api/python/lance.html#lance.dataset.LanceDataset.to_table) should be sufficient.

I was asking that the Dataset scan still works (i.e., returning empty table)

Comment (trueutkarsh) [reply rust/src/dataset.rs:1087]: Thanks for clarifying. I've added a test in python package where we check to_table() on empty dataset see the data is empty but schema is in place, then append a new record and verify the new state of the new dataset as well.

File: rust/src/arrow/record_batch.rs [hunk 4]
```diff
@@ -18,26 +18,35 @@
...
 use arrow_schema::{ArrowError, SchemaRef};
 
-use crate::Result;
+use crate::{Error, Result};
 
 /// RecordBatchBuffer is a in-memory buffer for multiple [`RecordBatch`]s.
 ///
 ///
 #[derive(Debug)]
 pub struct RecordBatchBuffer {
```
Comment (wjones127) [inline rust/src/arrow/record_batch.rs:27]: I'm wondering if we even want to keep this. It seems like all the uses can be handled either with `Vec<RecordBatch>` or `RecordBatchIterator<Vec<RecordBatch>::Iter>`. What do we think of removing it?

https://docs.rs/arrow-array/42.0.0/arrow_array/struct.RecordBatchIterator.html

Comment (trueutkarsh) [reply rust/src/arrow/record_batch.rs:27]: RecordBatchBuffer provides the ability to extend the record batches. RecordBatchIterator being an iterator (or wrapping an iterator) will only provide the read only view to the batches hence cannot do that. This functionality is used while writing batches limited by max rows per group in (dataset.rs::402, fragment.rs::84).
Vec<RecordBatch> provides that functionality but we cannot store schema for empty vector then.
Hence I think RecordBatchBuffer is best of both worlds so removing it is not feasible unless we change the implementation of fragment.rs::create and dataset.rs::write.
Please correct me if I misinterpreted anything. 
Thanks

Comment (wjones127) [reply rust/src/arrow/record_batch.rs:27]: But are there any place where we are adding batches at the same time as we are iterating?

If we are adding batches, I think we can just use `Vec<RecordBatch>`.
If we are iterating, we could just use `RecordBatchIterator<>`.

Is there a place where we need both at the same time?

Comment (trueutkarsh) [reply rust/src/arrow/record_batch.rs:27]: Well, you're right, thanks for pointing this out. After skimming through the critical places it doesn't seem like we need both functionality at the same time. I'll start making the changes accordingly.

File: rust/src/dataset.rs [hunk 5]
```diff
@@ -41,7 +41,7 @@ mod write;
 use self::feature_flags::{apply_feature_flags, can_read_dataset, can_write_dataset};
 use self::fragment::FileFragment;
 use self::scanner::Scanner;
-use crate::arrow::*;
+// use crate::arrow::*;
```
Comment (wjones127) [inline rust/src/dataset.rs:44]: Can we remove this?

**Review by wjones127 — COMMENTED** [review]
Looks pretty good. Just one comment.

**Review by wjones127 — APPROVED** [review]

**Review by eddyxu — APPROVED** [review]

**Pull request merged**